Cpp/exercises/numbers

AlexLeSang opened the pull request.

AlexLeSang committed:
[C++] Add .docs/introduction.md

Add introduction to integer and floating point numbers

AlexLeSang committed:
[C++] Add .docs/instructions.md

Describe the assembly speed exercise

AlexLeSang committed:
[C++] Add numbers core exercise

Add track-specific files:
- stub implementation files
- C++ project file
- the test suite
- an example implementation that passes all tests
- catch2 header
- generated main from test library

AlexLeSang committed:
[C++] Add numbers core exercise

Update author in .meta/config.json

AlexLeSang committed:
[C++] Add numbers core exercise

Extend list of implemented exercises

AlexLeSang committed:
[C++] Add numbers core exercise

Add numbers concept exercise entry the track's config.json

AlexLeSang committed:
[C++] Add numbers core exercise

Describe design

AlexLeSang committed:
[C++] Add numbers core exercise

Describe additional topics in .docs/after.md

AlexLeSang committed:
[C++] Add numbers core exercise

Provide hints

AlexLeSang committed:
[C++] Add numbers core exercise

Fix typos

AlexLeSang committed:
[C++] Add numbers core exercise

Remove catch.hpp because it violates 1MB file size limit

AlexLeSang committed:
[C++] Add numbers core exercise

Format files

siebenschlaefer commented on languages/cpp/exercises/concept/numbers/.docs/introduction.md:
```diff
@@ -0,0 +1,73 @@
+
+### Arithmetic operators
+
+Arithmetic is done using the standard [arithmetic operators][arithmetic-operators] (`+`, `-`, `*`, etc.). Numbers can be compared using the standard [comparison and equality operators][comparison-operators] (`<`, `>=`, etc.) and the equality-operator (`==`) and inequality operator (`!=`).
```
Since you list `==` and `!=` separately you could replace "comparison and equality operators" with "comparison operators".

siebenschlaefer commented on languages/cpp/exercises/concept/numbers/.docs/introduction.md:
```diff
@@ -0,0 +1,73 @@
+
+### Fixed width integer types
+
+By including `<cstdint>`, programmer gains access to [integer types with fixed width][fixed-width-integer-types]. This provides data-model independent way of representing integer numbers and ensures consistent results of arithmetic operations across all platforms.
```
Missing articles:
"the programmer" or "a programmer"
"provides a data-model"
But (disclaimer:) I'm not a native English speaker.

siebenschlaefer commented on languages/cpp/exercises/concept/numbers/.docs/introduction.md:
```diff
@@ -0,0 +1,73 @@
+
+There are special values for floating point types, such as:
+
+- infinity (positive and negative), see INFINITY
```
markup: \`INFINITY\`

siebenschlaefer commented on languages/cpp/exercises/concept/numbers/.docs/introduction.md:
````diff
@@ -0,0 +1,73 @@
+float f = 12e4;
+double d = 55.85;
+```
+
````
Shouldn't we mention suffixes?

AlexLeSang replied:
Do you mean binary, octal and hex?

AeroStun replied:
These would be prefixes
I believe Matthias wants to explain what `u` `l`, `ll`, `ull`, and `f` change to an integer literal or a double literal respectively

siebenschlaefer commented on languages/cpp/exercises/concept/numbers/.docs/introduction.md:
```diff
@@ -0,0 +1,73 @@
+
+Implicit conversions are defined between floating types and integer types.
+
+Compilation flags, such as `-Warith-conversion`, `-Wconversio`, `-Wfloat-conversion` and `-Wsign-conversion` for GCC/Clang, warn about possible unsafe numerical type conversions.
```
typo: `-Wconversion`

siebenschlaefer commented on languages/cpp/exercises/concept/numbers/.meta/design.md:
```diff
@@ -0,0 +1,26 @@
+- Know the different types of integer and floating point numbers.
+- Know integer type modifiers.
+- Know of basic operators such as multiplication, comparison and equality.
+- Know how to conditionally execute code using an `if` statement.
```
I didn't see anything about `if` statements. Did I miss it?

AlexLeSang replied:
The simplest way to do the rate map is with an `if` statement. Should I add this to references?

ErikSchierboom replied:
> I didn't see anything about if statements. Did I miss it?

Great observation @siebenschlaefer! One of the concepts this exercise introduces is conditional execution of code using if-statements. As such, the introduction.md (and after.md) should have a section on `if` statements.

siebenschlaefer commented on languages/cpp/exercises/concept/numbers/.meta/design.md:
```diff
@@ -0,0 +1,26 @@
+- `numbers`: know of the existence of the two most commonly used number types: integer and floating point; know of basic operators such as multiplication, comparison and equality.
+- `conditionals`: know how to conditionally execute code using an `if` statement.
+
+## Perquisites
```
typo: Prerequisites

siebenschlaefer commented on languages/cpp/exercises/concept/numbers/.meta/example.cpp:
```diff
@@ -0,0 +1,33 @@
+
+namespace assembly_line {
+
+constexpr uint32_t PRODUCTION_RATE_DEFAULT_SPEED = 221;
```
`<cstdint>` defines `std::uint32_t`, `<stdint.h>` defines `::uint32_t`.
Please signed types for arithmetic operations, see [ES.100-102 of the C++ Core Guidelines](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es100-dont-mix-signed-and-unsigned-arithmetic).
[ES.9 of the C++ Core Guidelines](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es9-avoid-all_caps-names) says: "*Avoid ALL_CAPS names*".

siebenschlaefer commented on languages/cpp/exercises/concept/numbers/numbers.h:
```diff
@@ -0,0 +1,6 @@
+#ifndef NUMBERS_H
+#define NUMBERS_H
+
+namespace log_line {}
```
namespace assembly_line {}

siebenschlaefer commented on languages/cpp/exercises/concept/numbers/.meta/example.cpp:
```diff
@@ -0,0 +1,33 @@
+    return PRODUCTION_RATE_DEFAULT_SPEED * speed * success_rate(speed);
+}
+
+int32_t working_items_per_minute(const int32_t speed) {
```
std::int32_t

siebenschlaefer commented on languages/cpp/exercises/concept/numbers/.meta/example.cpp:
```diff
@@ -0,0 +1,33 @@
+}
+}  // namespace
+
+double production_rate_per_hour(const int32_t speed) {
```
std::int32_t

siebenschlaefer commented on languages/cpp/exercises/concept/numbers/.meta/example.h:
```diff
@@ -0,0 +1,11 @@
+#include <cstdint>
+
+namespace assembly_line {
+double production_rate_per_hour(const int32_t speed);
```
std::int32_t

siebenschlaefer commented on languages/cpp/exercises/concept/numbers/.meta/example.h:
```diff
@@ -0,0 +1,11 @@
+
+namespace assembly_line {
+double production_rate_per_hour(const int32_t speed);
+int32_t working_items_per_minute(const int32_t speed);
```
std::int32_t

siebenschlaefer commented on languages/cpp/exercises/concept/numbers/numbers_test.cpp:
```diff
@@ -0,0 +1,68 @@
+
+TEST_CASE("Working items per minute for speed zero") {
+    auto rate = assembly_line::working_items_per_minute(0);
+    static_assert(std::is_same<int32_t, decltype(rate)>::value,
```
std::int32_t (also in the following tests)

siebenschlaefer commented on languages/cpp/exercises/concept/numbers/numbers_test.cpp:
```diff
@@ -0,0 +1,68 @@
+    auto rate = assembly_line::production_rate_per_hour(1);
+    static_assert(std::is_same<double, decltype(rate)>::value,
+                  "Returned value should be of type double");
+    REQUIRE(221.0 == rate);
```
Should we really compare exact values? I'd suggest
REQUIRE(Approx(221.0).margin(0.001) == rate);

AlexLeSang replied:
That is a very good point! Indeed, exact comparison of floats is generally bad idea. I'll add a margin.

siebenschlaefer commented on languages/cpp/exercises/concept/numbers/.docs/introduction.md:
```diff
@@ -0,0 +1,73 @@
+There are two different types of numbers in C++: integers and floats.
```
I'd prefer "floating point types"

siebenschlaefer commented on languages/cpp/exercises/concept/numbers/.docs/introduction.md:
```diff
@@ -0,0 +1,73 @@
+
+## Integer numbers
+
+The keyword to define an integer type is `int`. The keyword `int` can be omitted if one of the length modifiers is present. The actual width in bits depends on the data model (computer architecture).
```
"integer type" -> "integer variable"

siebenschlaefer reviewed (commented):
I like it but IMHO there are some rough edges.

AlexLeSang commented on languages/cpp/exercises/concept/numbers/.docs/after.md:
```diff
@@ -0,0 +1,52 @@
+### NULL
+
+[NULL][null] is a constant expression of integer type that evaluates to zero.
+
```
Constant `nullptr` is of type `std::nullptr_t`, while `NULL` is integer.

Do you think that there should be a mentioning about that in this context?

ErikSchierboom replied:
I would be hesitant to introduce pointers here. That seems like they should be introduced in a separate exercise.

AeroStun replied:
I would not introduce a C++ programmer to `NULL`, given it is part of the backwards-compat part of the language.
I think we should strive to first introduce the language as it is today, ignoring the legacy stuff, and then to provide an advanced exercise introducing old stuff so people can read legacy code and understand its behavior.
Besides, `#define NULL nullptr` is a valid C++ implementation of the `NULL` macro (see https://eel.is/c++draft/conv.ptr#1.sentence-1)

AlexLeSang replied:
OK, will remove.

AlexLeSang commented on languages/cpp/exercises/concept/numbers/.docs/after.md:
```diff
@@ -0,0 +1,52 @@
+
+### `std::is_arithmetic`
+
+If the type is arithmetic can be done with a meta-function [`std::is_arithmetic`][is_arithmetic].
```
I agree, this topic is too advance in the context of this exercise. I will remove it.

AlexLeSang committed:
[C++] Add numbers core exercise

Apply review comments:
 - remove std::is_arithmetic
 - remove integer semantic for char types
 - int32_t -> std::int32_t, in examples, definitions and tests
 - typos
 - rename constant to avoid all caps name
 - remove mixing of signed and unsigned arithmetic
 - fix incorrect namespace in the header file
 - use approximate comparison for floating point results

AlexLeSang commented on languages/cpp/exercises/concept/numbers/.docs/after.md:
```diff
@@ -0,0 +1,52 @@
+
+### Char types as integers
+
+Char types `char` `unsigned char`, `wchar_t`, `char16_t` and `char32_t` have the same memory layout and in some cases can be treated as integers.
```
Indeed, your point is correct. I'll remove the section.

Answering 2, I have never seen `signed char` used anywhere, but `char` used for storing 'small' integers I have seen a lot.

Answering 3, this is also connected to using `char` as numeric type and doing arithmetic on it. It works because `char` is an integer type but it is not obvious in the semantic of character.

ErikSchierboom commented on languages/cpp/exercises/concept/numbers/.meta/example.h:
```diff
@@ -0,0 +1,11 @@
+
+#include <cstdint>
+
+namespace assembly_line {
+double production_rate_per_hour(const std::int32_t speed);
+std::int32_t working_items_per_minute(const std::int32_t speed);
+}  // namespace assembly_line
```
Maybe add a little indentation here?

```suggestion
namespace assembly_line {
    double production_rate_per_hour(const std::int32_t speed);
    std::int32_t working_items_per_minute(const std::int32_t speed);
}
```

ErikSchierboom replied:
The example implementation file uses the `const` modifier. As per [the docs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#example-implementation-file), the example implementation should only use language features introduced by the exercise or its prerequisites (and their prerequisites, and so on). Is `const` introduced in one of the prerequisites?

AeroStun replied:
Besides, `const` value parameters should definitely not appear in declarations, as they don't do anything but confuse the reader.

Regarding indentation, we should probably all agree on a .clang-format document and stick to it; personally I never indent on namespaces

ErikSchierboom replied:
I'm fine with whatever convention is used, as long as the same convention is used for all exercises.

AlexLeSang replied:
> Maybe add a little indentation here?

I have used the same '.clang-format' file as the one in the cpp repository with exercises to format this exercise.

ErikSchierboom commented on languages/cpp/exercises/concept/numbers/numbers.cpp:
```diff
@@ -0,0 +1,3 @@
+#include "numbers.h"
+
+namespace assembly_line {}
```
The stub file does not contain stubs for the methods the student needs to implement. These should be added (see [the docs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#stub-implementation-file)).

AlexLeSang replied:
Added

ErikSchierboom commented on languages/cpp/exercises/concept/numbers/numbers.h:
```diff
@@ -0,0 +1,6 @@
+#ifndef NUMBERS_H
+#define NUMBERS_H
+
+namespace assembly_line {}
```
The stub file does not contain stubs for the methods the student needs to implement. These should be added (see [the docs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#stub-implementation-file)).

AlexLeSang replied:
Added

ErikSchierboom commented on languages/cpp/exercises/concept/numbers/.docs/after.md:
```diff
@@ -0,0 +1,43 @@
+
+### Integer literals
+
+Integer values can have different written form - "literal" that helps create integer numbers in different notations.
```
```suggestion
Integer values can be written using several different notations.
```

AlexLeSang replied:
Thanks!

ErikSchierboom commented on languages/cpp/exercises/concept/numbers/.docs/hints.md:
```diff
@@ -0,0 +1,12 @@
+## General
+
+- Determining the success rate can be done through a [conditional statement][if-statement].
+- C++ allows for multiplication to be applied to two different number types (such as an `int` and a `double`). It will automatically return the "largest" data type.
+- Ranges of integer and floating point types listed in [fundamental types][fundamental-types]
+- Fixed-width integer types, such as `std::int32_t`, are listed in [fixed width integer types][fixed-width-integer-types].
+- Information about types and how to acquire it during compilation is available in [numeric limits library][numeric-limits].
```
The hints are nice! Are there hints that can be linked to a specific task? See [this file](https://github.com/exercism/v3/blob/master/languages/cpp/exercises/concept/strings/.docs/hints.md) for an example on how to link specific hints to specific tasks.

ErikSchierboom commented on languages/cpp/exercises/concept/numbers/.docs/after.md:
```diff
@@ -0,0 +1,43 @@
+### Special integer types
+
+There are two other important integer types often used in the standard library:
```
It looks like the `after.md` document adds to what is listed in the `introduction.md` document. As per [the docs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docsaftermd), this file should at least contain all information that is introduced in the `introduction.md` document.

ErikSchierboom commented on languages/cpp/exercises/concept/numbers/.docs/introduction.md:
```diff
@@ -0,0 +1,73 @@
+There are two different types of numbers in C++: integers and floating point types.
```
This introduction has got tons of information. The information is great, but we strive to keep the introduction.md as short as can be (see [the docs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docsintroductionmd). To keep it as short as possible, the following guidelines can help:

- The information provided should give the student just enough context to figure out the solution themselves.
- Only information that is needed to understand the fundamentals of the concept and solve the exercise should be provided. Extra information should be left for the after.md.

Could you take a second look at the information in this introduction.md and look to shorten it a bit?

AeroStun replied:
IMO, for the purpose of self-learning, we should mention that numeric types in C++ are called "arithmetic types"; lets the student search more accurately for extra information

ErikSchierboom replied:
> lets the student search more accurately for extra information

This is indeed a very important point! It's even in [the docs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docsintroductionmd):

> Proper technical terms should be used so that the student can easily search for more information.

ErikSchierboom commented on languages/cpp/exercises/concept/numbers/.meta/example.cpp:
```diff
@@ -0,0 +1,33 @@
+
+namespace assembly_line {
+
+constexpr std::int32_t production_rate_default_speed = 221;
```
Are students familiar with `constexpr` at this point? As per the docs, the example implementation should only use language features introduced by the exercise or its prerequisites (and their prerequisites, and so on).

ErikSchierboom commented on languages/cpp/exercises/concept/numbers/.meta/design.md:
```diff
@@ -0,0 +1,26 @@
+
+## Prerequisites
+
+There are no prerequisites.
```
I think the `basics` concept is a prerequisite.

ErikSchierboom reviewed (commented):
Thanks for working on this! I've left some comments.

AlexLeSang committed:
Merge remote-tracking branch 'upstream/master' into cpp/exercises/numbers

AlexLeSang committed:
[C++] Add numbers core exercise

Apply review comments:
 - remove mentions of `NULL`

AlexLeSang committed:
[C++] Add numbers core exercise

Apply review comments:
 - remove `const` from declarations and definitions

AlexLeSang committed:
[C++] Add numbers core exercise

Apply review comments:
  - add stubs according to https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#stub-implementation-file

AlexLeSang committed:
[C++] Add numbers core exercise

Apply review comments:
  - Shorten the introduction

AlexLeSang committed:
[C++] Add numbers core exercise

Apply review comments:
  - Simplify introduction
  - Simplify the after section
  - Add reference to arithmetic types

ErikSchierboom commented:
@AlexLeSang Let me know when you'd like to receive a new review.

ErikSchierboom commented:
This Friday (Jan 29th 2021), we're going to explode this repo back into the individual track repositories and archive this repo.

As such, this PR needs to be merged by Friday 9am UTC, or it will need to be manually reopened against the track repo.
